### ISSUE TITLE
readers.las: multiple file opens for info

### ISSUE THREAD

**Issue by rcoup** [PDAL/PDAL#3846]
**Describe the bug**

Initially I was investigating HTTPS/S3 access to (non-COPC) LAS/LAZ files, and realised `pdal info` would end up doing two complete file downloads.

```
$ VERBOSE=1 build/bin/pdal -v 8 info --schema --metadata --driver=readers.las "https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz"
...
> GET /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> HEAD /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> GET /media/PDAL/data/master/autzen/autzen.laz HTTP/2
> HEAD /media/PDAL/data/master/autzen/autzen.laz HTTP/2
```

So, readers.las doesn't have the ability to do partial range reads over HTTP, it ends up downloading the whole file. But that's a separate issue. 

The duplicate file opens happen locally too:
```
$ build/bin/pdal -v 8 info --schema --metadata --driver=readers.las test/data/autzen/autzen.las
$ sudo dtruss -n pdal 2>&1 | grep -P "(open.*autzen)|(close.*?\(0x4)|(read.*?\(0x4)"
 9940/0xd94b:  open_nocancel("test/data/autzen/autzen.las\0", 0x0, 0x0)		 = 4 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "\314\003\334\343\023\005\364\242\0", 0x1000)		 = 866 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "\314\003\334\343\023\005\364\242\0", 0x1000)		 = 866 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "PROJCS[\"NAD83(HARN) / Oregon Lambert (ft)\",GEOGCS[\"NAD83(HARN)\",DATUM[\"NAD83_High_Accuracy_Regional_Network\",SPHEROID[\"GRS 1980\",6378137,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY[\"EPSG\",\"6152\"]],PRIMEM[\"Greenwich\",0,AUTHORITY", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "\001\0", 0x1000)		 = 3907 0
 9940/0xd94b:  read_nocancel(0x4, "false_easting\",1312335.958],PARAMETER[\"false_northing\",0],AUTHORITY[\"EPSG\",\"2994\"],AXIS[\"X\",EAST],AXIS[\"Y\",NORTH]]\0", 0x1000)		 = 0 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "NAD83(HARN) / Oregon Lambert (ft)|NAD83(HARN)|\0", 0x1000)		 = 3789 0
 9940/0xd94b:  read_nocancel(0x4, "\003Y\024\020\005\037\237\0", 0x1000)		 = 0 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  read_nocancel(0x4, "PROJCS[\"NAD83(HARN) / Oregon Lambert (ft)\",GEOGCS[\"NAD83(HARN)\",DATUM[\"NAD83_High_Accuracy_Regional_Network\",SPHEROID[\"GRS 1980\",6378137,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY[\"EPSG\",\"6152\"]],PRIMEM[\"Greenwich\",0,AUTHORITY", 0x1000)		 = 3688 0
 9940/0xd94b:  read_nocancel(0x4, "d\024\213Z~\n\016A\035'\313\003\021\354\020\005\033\246\0", 0x1000)		 = 0 0
 9940/0xd94b:  close_nocancel(0x4)		 = 0 0
 9940/0xd94b:  open_nocancel("test/data/autzen/autzen.las\0", 0x0, 0x0)		 = 4 0
 9940/0xd94b:  read_nocancel(0x4, "LASF\0", 0x1000)		 = 4096 0
 9940/0xd94b:  close_nocancel(0x4)		 = 0 0
 9940/0xd94b:  read_nocancel(0x4, "TZif2\0", 0xA1E8)		 = 3661 0
 9940/0xd94b:  close_nocancel(0x4)		 = 0 0
 9940/0xd94b:  read_nocancel(0x4, "TZif2\0", 0xA1E8)		 = 3519 0
 9940/0xd94b:  close_nocancel(0x4)		 = 0 0
```

Which explains the duplicate downloads when going via Arbiter/http. Backtraces of the `openFile()` calls:
```
FIRST
  * frame #0: 0x0000000105aaddc8 libpdalcpp.15.dylib`pdal::Utils::openFile(path="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz", asBinary=true) at PDALUtils.cpp:313:15
    frame #1: 0x00000001057a3e50 libpdalcpp.15.dylib`pdal::LasReader::LasStreamIf::LasStreamIf(this=0x0000600000010010, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at LasReader.hpp:72:27
    frame #2: 0x0000000105798ef4 libpdalcpp.15.dylib`pdal::LasReader::LasStreamIf::LasStreamIf(this=0x0000600000010010, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at LasReader.hpp:72:13
    frame #3: 0x0000000105798c58 libpdalcpp.15.dylib`pdal::LasReader::createStream(this=0x0000000104104080) at LasReader.cpp:188:26
    frame #4: 0x0000000105799214 libpdalcpp.15.dylib`pdal::LasReader::initializeLocal(this=0x0000000104104080, table=0x0000600003e00300, m=0x0000000104104170) at LasReader.cpp:215:5
    frame #5: 0x0000000105798634 libpdalcpp.15.dylib`pdal::LasReader::initialize(this=0x0000000104104080, table=0x0000600003e00300) at LasReader.cpp:157:5
    frame #6: 0x0000000105b6ab28 libpdalcpp.15.dylib`pdal::Stage::prepare(this=0x0000000104104080, table=0x0000600003e00300) at Stage.cpp:168:5
    frame #7: 0x0000000105b6aae0 libpdalcpp.15.dylib`pdal::Stage::prepare(this=0x0000000104104530, table=0x0000600003e00300) at Stage.cpp:164:15
    frame #8: 0x0000000105abf898 libpdalcpp.15.dylib`pdal::PipelineManager::execute(this=0x0000000104004b58, mode=PreferStream) at PipelineManager.cpp:245:12
    frame #9: 0x000000010599c004 libpdalcpp.15.dylib`pdal::InfoKernel::run(this=0x0000000104004b40, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at InfoKernel.cpp:250:23
    frame #10: 0x000000010599d478 libpdalcpp.15.dylib`pdal::InfoKernel::execute(this=0x0000000104004b40) at InfoKernel.cpp:323:25
    frame #11: 0x0000000105a7b550 libpdalcpp.15.dylib`pdal::Kernel::innerRun(this=0x0000000104004b40, args=0x000000016fdfe598) at Kernel.cpp:283:12
    frame #12: 0x0000000105a7b268 libpdalcpp.15.dylib`pdal::Kernel::doExecution(this=0x0000000104004b40, args=0x000000016fdfe598) at Kernel.cpp:247:18
    frame #13: 0x0000000105a7b17c libpdalcpp.15.dylib`pdal::Kernel::run(this=0x0000000104004b40, cmdArgs=size=4, log=std::__1::shared_ptr<pdal::Log>::element_type @ 0x00000001040049d0 strong=8 weak=1) at Kernel.cpp:235:12
    frame #14: 0x0000000100007e34 pdal`App::execute(this=0x000000016fdfe8a0, cmdArgs=size=4, log=std::__1::shared_ptr<pdal::Log>::element_type @ 0x00000001040049d0 strong=8 weak=1) at pdal.cpp:371:27
    frame #15: 0x0000000100007974 pdal`main(argc=8, argv=0x000000016fdfeaa8) at pdal.cpp:315:17
    frame #16: 0x000000010018d08c dyld`start + 520

SECOND
  * frame #0: 0x0000000105aaddc8 libpdalcpp.15.dylib`pdal::Utils::openFile(path="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz", asBinary=true) at PDALUtils.cpp:313:15
    frame #1: 0x00000001057a3e50 libpdalcpp.15.dylib`pdal::LasReader::LasStreamIf::LasStreamIf(this=0x0000600000006a50, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at LasReader.hpp:72:27
    frame #2: 0x0000000105798ef4 libpdalcpp.15.dylib`pdal::LasReader::LasStreamIf::LasStreamIf(this=0x0000600000006a50, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at LasReader.hpp:72:13
    frame #3: 0x0000000105798c58 libpdalcpp.15.dylib`pdal::LasReader::createStream(this=0x0000000104104080) at LasReader.cpp:188:26
    frame #4: 0x000000010579af54 libpdalcpp.15.dylib`pdal::LasReader::ready(this=0x0000000104104080, table=0x0000600003e00300) at LasReader.cpp:336:5
    frame #5: 0x0000000105b9935c libpdalcpp.15.dylib`pdal::Streamable::execute(this=0x000000016fdfdd18, table=0x0000600003e00300)::StreamableList::ready(pdal::BasePointTable&) at Streamable.cpp:100:20
    frame #6: 0x0000000105b98af0 libpdalcpp.15.dylib`pdal::Streamable::execute(this=0x0000000104104530, table=0x0000600003e00300) at Streamable.cpp:155:38
    frame #7: 0x0000000105abf8d0 libpdalcpp.15.dylib`pdal::PipelineManager::execute(this=0x0000000104004b58, mode=PreferStream) at PipelineManager.cpp:255:12
    frame #8: 0x000000010599c004 libpdalcpp.15.dylib`pdal::InfoKernel::run(this=0x0000000104004b40, filename="https://media.githubusercontent.com/media/PDAL/data/master/autzen/autzen.laz") at InfoKernel.cpp:250:23
    frame #9: 0x000000010599d478 libpdalcpp.15.dylib`pdal::InfoKernel::execute(this=0x0000000104004b40) at InfoKernel.cpp:323:25
    frame #10: 0x0000000105a7b550 libpdalcpp.15.dylib`pdal::Kernel::innerRun(this=0x0000000104004b40, args=0x000000016fdfe598) at Kernel.cpp:283:12
    frame #11: 0x0000000105a7b268 libpdalcpp.15.dylib`pdal::Kernel::doExecution(this=0x0000000104004b40, args=0x000000016fdfe598) at Kernel.cpp:247:18
    frame #12: 0x0000000105a7b17c libpdalcpp.15.dylib`pdal::Kernel::run(this=0x0000000104004b40, cmdArgs=size=4, log=std::__1::shared_ptr<pdal::Log>::element_type @ 0x00000001040049d0 strong=8 weak=1) at Kernel.cpp:235:12
    frame #13: 0x0000000100007e34 pdal`App::execute(this=0x000000016fdfe8a0, cmdArgs=size=4, log=std::__1::shared_ptr<pdal::Log>::element_type @ 0x00000001040049d0 strong=8 weak=1) at pdal.cpp:371:27
    frame #14: 0x0000000100007974 pdal`main(argc=8, argv=0x000000016fdfeaa8) at pdal.cpp:315:17
    frame #15: 0x000000010018d08c dyld`start + 520
```

which leads to `createStream()`:
https://github.com/PDAL/PDAL/blob/6cb54d87eb6c133aa1865401a6fe163be2181fc3/io/LasReader.cpp#L184-L196

The warning isn't logged there, but the `m_streamif` pointer is also reset at the end of `initializeLocal()`:

https://github.com/PDAL/PDAL/blob/6cb54d87eb6c133aa1865401a6fe163be2181fc3/io/LasReader.cpp#L330-L331

**Expected behavior**

`pdal info` performs a single LAS file open (and a single GET request over HTTP), not two.

A super-crude patch passes all the tests, but I'm unsure what the other implications of it are?

```patch
diff --git a/io/LasReader.cpp b/io/LasReader.cpp
index 85e40eaf3..68daf5c5f 100644
--- a/io/LasReader.cpp
+++ b/io/LasReader.cpp
@@ -183,9 +183,9 @@ QuickInfo LasReader::inspect()

 void LasReader::createStream()
 {
-    if (m_streamIf)
-        std::cerr << "Attempt to create stream twice!\n";
-    m_streamIf.reset(new LasStreamIf(m_filename));
+    if (!m_streamIf) {
+        m_streamIf.reset(new LasStreamIf(m_filename));
+    }
     if (!m_streamIf->m_istream)
     {
         std::ostringstream oss;
@@ -327,7 +327,7 @@ void LasReader::initializeLocal(PointTableRef table, MetadataNode& m)
     for (int i = 0; i < (int)d->vlrs.size(); ++i)
         las::addVlrMetadata(d->vlrs[i], "vlr_" + std::to_string(i), forward, m);
-
-    m_streamIf.reset();
 }
```

**System/installation information:**

```
$ uname -a
Darwin inchgarvie.local 21.6.0 Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000 arm64 arm Darwin

$ pdal --version
pdal 2.5.0 (git-version: 6cb54d)
```

**Comment by abellgithub** [comment]
I'm not sure what the actual issue is. It would be good if you could explain that. Is the concern that the source file is opened more than once during some particular operation? If so, this is by design and is important to avoid errors with too many open file descriptors on some systems.

If the concern is the a file is downloaded more than once using a remote URL, that's a separate issue.

**Comment by rcoup** [comment]
> I'm not sure what the actual issue is. It would be good if you could explain that. Is the concern that the source file is opened more than once during some particular operation? If so, this is by design and is important to avoid errors with too many open file descriptors on some systems.
>
> If the concern is the a file is downloaded more than once using a remote URL, that's a separate issue.

Running a pipeline on a LAS file shouldn't download a potentially large file twice from the network, that seems like a clear bug. And (IME) given the expense of metadata operations on distributed filesystems (Gluster/NFS/etc) I'd suggest two `open()` calls isn't great in that sort of storage scenario either.

But yes, default file descriptor limits can be annoyingly low. If you think the open-it-twice behaviour is good then I guess I can focus on arbiter/FileUtils and build up some sort of hint/cache/reuse approach there instead.

**Comment by abellgithub** [comment]
I would suggest closing this ticket and creating a new one that clearly expresses a single issue. Debugging information and an attempt to solve the problem within the description itself makes things hard to follow (at least it does for me).

**Comment by hobu** [comment]
Competing desires with this proposed change:

* Don't want to run out of file handles for things like `pdal tile`
* Don't want to download content extra times because the fetch-on-first-access happens repeatedly due to Arbiter file handle going away.

If we were to apply this, our users would be more likely to have to figure out how to mess with setting their `ulimit` open file count. Maybe that isn't such a bad thing, IDK. 

I would rather do this than #3847 where we would put a caching layer inside of Arbiter, but I question whether or not we should put another option in `readers.las` to control this behavior.

**Comment by rcoup** [comment]
My comment/research on #3857

> Hard FD limits are circa 1M descriptors on most Linux distros by default, unlimited on macOS, 16M on Windows. That goes a long way...

**Comment by abellgithub** [comment]
This is a real problem for Windows. Just raising the number of open FDs
isn't a good option.

**Comment by hobu** [comment]
Were we running out of them before? I can't quite remember what the story was with them, but I'm pretty sure that we opened and closed the file because of `pdal tile` sucking up the handles.

**Comment by abellgithub** [comment]
Yes. Also, users will do crazy stuff (we provide an API). I honestly think the best way to deal is to let users download stuff on their own, but I think that ship has already sailed. We're kind of teasing in that it seems as if we're doing smart stuff across the network, but except for the cloud-based file types, we're not.

My guess is that this really comes up only with LAS/LAZ, so maybe it's worth some work to deal with remote LAS/LAZ differently. You don't much want to download a big LAZ if all you want is header info.
